tests/ts_updown_util.erl

```diff
@@ -0,0 +1,372 @@
+
+setup_cluster(Config) ->
+    %% build the starting (old = upgraded, current) cluster
+    Nodes = rt:build_cluster(
+              lists:duplicate(3, current)),
+    Config ++
+        [
+         {nodes, Nodes}
+        ].
+
+setup_client(Config) ->
+    %% By default, we use the client in the 'current' version for all
+    %% queries.  Add `{use_previous_client, true}` to the Config arg
+    %% when calling it from your_module:init_per_suite to change that
+    %% to selectively use old code to connect to downgraded nodes.
+    UsePreviousClient = proplists:get_value(use_previous_client, Config, false),
+    PrevClientNode = maybe_setup_slave_for_previous_client(UsePreviousClient),
+    Config ++
+        [
+         {previous_client_node, PrevClientNode}
+        ].
+
+maybe_setup_slave_for_previous_client(true) ->
+    %% set up a separate, slave node for the 'previous' version
+    %% client, to talk to downgraded nodes
+    _ = application:start(crypto),
+    Suffix = [crypto:rand_uniform($a, $z) || _ <- [x,x,x,i,x,x,x,i]],
+    PrevRiakcNode = list_to_atom("alsoran_"++Suffix++"@127.0.0.1"),
+    rt_client:set_up_slave_for_previous_client(PrevRiakcNode);
+maybe_setup_slave_for_previous_client(_) ->
+    node().
+
+maybe_shutdown_client_node(Config) ->
+    case ?CFG(previous_client_node, Config) of
+        ThisNode when ThisNode == node() ->
+            ok;
+        SlaveNode ->
+            rt_slave:stop(SlaveNode)
+    end.
+
+
+%% scenarios
+
+-spec run_scenarios(config(), [#scenario{}]) -> [#failure_report{}].
+run_scenarios(Config, Scenarios) ->
+    Failures =
+        lists:foldl(
+          fun(Scenario, FF) ->
+                  run_scenario(Config, Scenario) ++ FF
+          end,
+          [], Scenarios),
+
+    Failures.
+
+
+-spec run_scenario(config(), #scenario{})
+                  -> [#failure_report{}].
+run_scenario(Config,
+             #scenario{table_node_vsn = TableNodeVsn,
+                       query_node_vsn = QueryNodeVsn,
+                       need_table_node_transition = NeedTableNodeTransition,
+                       need_query_node_transition = NeedQueryNodeTransition,
+                       need_pre_cluster_mixed = NeedPreClusterMixed,
+                       need_post_cluster_mixed = NeedPostClusterMixed,
+                       %% for these, we may have invariants in Config:
+                       data = Data_,
+                       ddl = DDL_,
+                       select_vs_expected = SelectVsExpected_}) ->
+    NodesAtVersions0 =
+        [{N, rtdev:node_version(rtdev:node_id(N))} || N <- ?CFG(nodes, Config)],
+
+    %% 0. retreive scenario-invariant data from Config
+    [Data, DDLFmt, SelectVsExpected] =
+        [begin
+             case Supplied of
+                 undefined ->
+                     ?CFG(Item, Config);
+                 Defined ->
+                     Defined
+             end
+         end || {Supplied, Item} <- [{Data_, data}, {DDL_, ddl},
```

Comment:
Well, they are not proplists but short-lived aggregates to be immediately decomposed within this list comprehension, but I can't help agreeing they look.. unnatural. Fixed in ba73dca.